rust_src/src/dired_rust.rs

```diff
@@ -0,0 +1,572 @@
+///   If even HIGH is too large for an Emacs integer, this is instead of the form
+///   (HIGH MIDDLE . LOW): first the high bits, then the middle 24 bits,
+///   and finally the low 16 bits.
+/// 11. Filesystem device number.  If it is larger than what the Emacs
+///   integer can hold, this is a cons cell, similar to the inode number.
+///
+/// On most filesystems, the combination of the inode and the device
+/// number uniquely identifies the file.
+///
+/// On MS-Windows, performance depends on `w32-get-true-file-attributes',
+/// which see.
+///
+/// On some FAT-based filesystems, only the date of last access is recorded,
+/// so last access time will always be midnight of that day.
+#[lisp_fn(min = "1")]
+fn file_attributes(filename: LispObject, id_format: LispObject) -> LispObject {
+    let fnexp_raw = unsafe { Fexpand_file_name(filename.to_raw(), Qnil) };
+    let fnexp = LispObject::from_raw(fnexp_raw);
+
+    let handler_raw = unsafe { Ffind_file_name_handler(fnexp_raw, Qfile_attributes) };
+    let handler = LispObject::from_raw(handler_raw);
+    if handler.is_not_nil() {
+        if id_format.is_not_nil() {
+            let res =
+                unsafe { call3(handler_raw, Qfile_attributes, fnexp_raw, id_format.to_raw()) };
+            return LispObject::from_raw(res);
+        } else {
+            let res = unsafe { call2(handler_raw, Qfile_attributes, fnexp_raw) };
+            return LispObject::from_raw(res);
+        }
+    }
+
+    file_attributes_internal(fnexp, id_format)
+}
+
+/// Return t if first arg file attributes list is less than second.
+/// Comparison is in lexicographic order and case is significant.  */)
+#[lisp_fn]
+fn file_attributes_lessp(f1: LispObject, f2: LispObject) -> LispObject {
+    LispObject::from_bool(string_ltp(car(f1), car(f2)))
+}
+
+/// Return a list of user names currently registered in the system.
+/// If we don't know how to determine that on this platform, just
+/// return a list with one element, taken from `user-real-login-name'.  */)
+#[lisp_fn]
+fn system_users() -> LispObject {
+    let mut done = false;
+    let mut unames = Vec::new();
+
+    while !done {
+        let pw: *mut libc::passwd = unsafe { libc::getpwent() };
+        if pw.is_null() {
+            done = true;
+        } else {
+            let pwnam_string =
+                unsafe { CStr::from_ptr((*pw).pw_name).to_string_lossy().into_owned() };
+            let pwnam_str = &*pwnam_string;
+            unames.push(str2lo(pwnam_str));
+        }
+    }
+    unsafe { libc::endpwent() };
+
+    if unames.len() < 1 {
+        unames.push(str2lo("rms"));
+    }
+
+    list(&mut unames)
+}
+
+fn string_ltp(a: LispObject, b: LispObject) -> bool {
+    let sa = lo2string(a);
+    let sb = lo2string(b);
+
+    if sa < sb { true } else { false }
+}
+
+fn string_gtp(a: LispObject, b: LispObject) -> bool {
+    let sa = lo2string(a);
+    let sb = lo2string(b);
+
+    if sa > sb { true } else { false }
```

Comment:
This return statement could be reduced to  just `sa > sb`

Reply:
Ok. Will update code.